sdk/storage/AzureStorageBlob/Source/Operation/BlockOperations.swift

```diff
@@ -45,17 +45,18 @@ internal class BlockOperation: ResumableOperation {
             return
         }
         let parent = transfer.parent
-        if isCancelled || isPaused { return }
+        if !transfer.isActive { return }
         transfer.state = .inProgress
         parent.state = .inProgress
         notifyDelegate(withTransfer: transfer)
-        if isCancelled || isPaused { return }
+        if !transfer.isActive { return }
         let group = DispatchGroup()
         group.enter()
         switch parent.transferType {
         case .download:
             guard let downloader = parent.downloader else {
                 assertionFailure("Downloader not found for BlobOperation.")
+                group.leave()
```

Comment:
Would it be a good idea to instead follow every call to `group.enter()` with a `defer { group.leave() }` ? That way we don't have to make sure we call `leave()` at every return point in the method, and it will also catch if the method is terminated by throwing an exception rather than returning.

Reply:
In theory it would be elegant and wonderful, but in practice, this results in an EXC_BAD_INSTRUCTION. It effectively does this:

```Swift
  group.enter()
  defer { group.leave() }
  {
    // Works great for early returns
  }
  group.wait()
  super.main()
  // fails here
}
```

Reply:
You have to call `group.leave()` in the normal scenario, resulting in a double call to `leave()` or the leave won't logically get called until after the `wait()` and thus the application will hang.

Reply:
I pushed a change that works. The key is to place the defer statement in the beginning of the callback closures. You can't just place one right after the intial enter call.